src/common/transformations/src/transformations/sdpa_to_paged_attention/total_sequence_length_pattern.cpp

```diff
@@ -23,21 +24,56 @@ ov::pass::TotalSequenceLengthPattern::TotalSequenceLengthPattern(
+    auto seq = pattern::wrap_type<v8::Gather>({kv_shape, gather_idx_label, pattern::any_input()});
 
     ov::matcher_pass_callback callback = [=](ov::pass::pattern::Matcher& m) {
         // TODO: Check that seq has axis that really takes sequence len but not any other dimension --
         //  use symbolic infra or look at the constant input
-        auto gather = m.get_match_root();
-        auto target_type = gather->get_output_element_type(0);
-        std::shared_ptr<Node> replacement = max_context_len;
-        if (replacement->get_output_element_type(0) != target_type) {
-            replacement = std::make_shared<v0::Convert>(replacement, target_type);
+        const auto& pattern_map = m.get_pattern_value_map();
+
+        auto concat = std::dynamic_pointer_cast<v0::Concat>(pattern_map.at(kv_concat).get_node_shared_ptr());
+        auto gather = std::dynamic_pointer_cast<v8::Gather>(pattern_map.at(seq).get_node_shared_ptr());
+        auto gather_idx =
+            std::dynamic_pointer_cast<v0::Constant>(pattern_map.at(gather_idx_label).get_node_shared_ptr());
+        auto gather_idx_data = *gather_idx->get_data_ptr<int64_t>();
+        if (!concat || !gather || !gather_idx) {
+            return false;
+        }
+
+        int64_t gather_idx_to_compare = gather_idx_data;
+
+        if (gather_idx_data < 0) {
+            if (gather->input(0).get_partial_shape().is_static()) {
+                const auto& gather_data_shape = gather->input(0).get_shape();
+                gather_idx_to_compare = ov::util::normalize(gather_idx_data, gather_data_shape[0]);
+            }
         }
-        auto required_shape = gather->get_output_partial_shape(0);
-        if (replacement->get_output_partial_shape(0) != required_shape && required_shape.rank().is_static()) {
-            replacement = op::util::reshapeTo(replacement, Shape(required_shape.rank().get_length(), 1));
+
+        std::shared_ptr<Node> replacement = max_context_len;
+
+        if (concat->get_axis() == gather_idx_to_compare) {
+            auto target_type = gather->get_output_element_type(0);
+
+            if (replacement->get_output_element_type(0) != target_type) {
+                replacement = std::make_shared<v0::Convert>(replacement, target_type);
+            }
+
+            auto required_shape = gather->get_output_partial_shape(0);
+
+            if (replacement->get_output_partial_shape(0) != required_shape && required_shape.rank().is_static()) {
+                replacement = op::util::reshapeTo(replacement, Shape(required_shape.rank().get_length(), 1));
+            }
+        } else {
+            if (concat->get_output_partial_shape(0)[gather_idx_data].is_static()) {
+                replacement =
+                    op::v0::Constant::create(element::i32,
+                                             Shape{},
+                                             {concat->get_output_partial_shape(0)[gather_idx_data].get_length()});
+            }
```

Comment:
At @72 I take the dimension the Gather would "cut-out". I take it from Concat because it is the only way I found to do it. Is there any way to take it from Gather, not from Concat as that would seem more natural for me?

Reply:
discussed offline

Reply:
Fixed.